Execution cleanup in integration tests

npauzenga opened the pull request:
**Summary:** 

This should delete executions from Postgres (and Dynamo) and update integration tests to use this new endpoint rather than calling the Dynamo Execution model directly.

Much of this will look eerily similar to https://github.com/nasa/cumulus/pull/2245

## PR Checklist

- [ ] Update CHANGELOG
- [x] Unit tests
- [ ] Adhoc testing
- [x] Integration tests

npauzenga commented on example/spec/parallel/ingest/ingestFromPdrSpec.js:
```diff
@@ -141,10 +142,10 @@ describe('Ingesting from PDR', () => {
     await Promise.all([
       deleteFolder(config.bucket, testDataFolder),
+      deleteExecution({ prefix: config.stackName, executionArn: workflowExecution.executionArn }),
+      deleteExecution({ prefix: config.stackName, executionArn: parsePdrExecutionArn }),
```
D'oh! These need to be outside of the Promise.all. Fixing now.

npauzenga commented:
I'm realizing that the integration tests changes I made here cover the tests where we were _trying_ to delete executions and failing but not cases where we weren't even trying. I'm going to be committing more of the same changes to other integration tests until I don't see errors in the CW logs.

npauzenga commented:
Added the cleanup to a bunch of tests that didn't have any executions cleanup: https://github.com/nasa/cumulus/pull/2252/commits/51528009e39374a2f733f2d17b5a0b0ff078d03e

botanical commented on example/spec/parallel/ingestGranule/IngestGranuleFailureSpec.js:
```diff
@@ -103,11 +105,11 @@ describe('The Ingest Granule failure workflow', () => {
 
   afterAll(async () => {
     // clean up stack state added by test
+    await deleteExecution({ prefix: config.stackName, executionArn: workflowExecution.executionArn });
```
Is there a reason you don't include this call in the Promise.all below? 🤔

npauzenga replied:
Yeah, that's exactly it. We need to be sure that executions are deleted before collections.

botanical commented on example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingErrorSpec.js:
```diff
@@ -204,6 +208,9 @@ describe('The IngestGranuleCatchDuplicateErrorTest workflow with DuplicateHandli
     ).catch(console.error);
 
+    await deleteExecution({ prefix: config.stackName, executionArn: firstIngestGranuleExecutionArn });
+    await deleteExecution({ prefix: config.stackName, executionArn: secondIngestGranuleExecutionArn });
```
🐜 I think you can combine these into a Promise.all but it's definitely a nit

botanical replied:
To save myself from copy 🍝 , this comment is applicable to other tests too

Jkovarik replied:
Oops, I made some similar comments below 😅         I did suggest promise.allSettled for some of them, but realize we haven't been doing a 'remove everything you can then fail' approach to the afterAlls, so a promise.all would be fine.

npauzenga replied:
Yup, in a lot of cases the order of deletes matters but not here. Updated a bunch of these to either use Promise.all or add a note explaining why not:

https://github.com/nasa/cumulus/pull/2252/commits/bf4c4749b365348d77fcaeedef7751611b732428

botanical commented on example/spec/parallel/testAPI/ruleSpec.js:
```diff
@@ -65,6 +68,7 @@ describe('When I create a scheduled rule via the Cumulus API', () => {
       prefix: config.stackName,
       ruleName: scheduledRuleName,
     });
+    await deleteExecution({ prefix: config.stackName, executionArn: executionArn });
```
🐜 
```suggestion
    await deleteExecution({ prefix: config.stackName, executionArn });
```

botanical commented on example/spec/parallel/testAPI/ruleSpec.js:
```diff
@@ -179,6 +187,8 @@ describe('When I create a scheduled rule with an executionNamePrefix via the Cum
       prefix: config.stackName,
       ruleName: scheduledRuleName,
     });
+    await deleteExecution({ prefix: config.stackName, executionArn: executionArn });
```
```suggestion
    await deleteExecution({ prefix: config.stackName, executionArn });
```

botanical commented on example/spec/parallel/testAPI/ruleSpec.js:
```diff
@@ -354,12 +377,13 @@ describe('When I create a one-time rule with an executionNamePrefix via the Cumu
 
   afterAll(async () => {
     console.log(`deleting rule ${helloWorldRule.name}`);
+    await deleteExecution({ prefix: config.stackName, executionArn: executionArn });
```
```suggestion
    await deleteExecution({ prefix: config.stackName, executionArn });
```

npauzenga commented on packages/api/endpoints/executions.js:
```diff
@@ -45,7 +46,50 @@ async function get(req, res) {
+    throw error;
+  }
+
+  try {
+    await knex.transaction(async (trx) => {
+      await executionPgModel.delete(trx, { arn });
+      await executionModel.delete({ arn });
+    });
+    return res.send({ message: 'Record deleted' });
+  } catch (error) {
+    // Delete is idempotent, so there may not be a DynamoDB
+    // record to recreate
+    if (existingExecution) {
+      await executionModel.create(existingExecution);
+    }
+    throw error;
+  }
+}
```
@botanical had a good question here along the lines of "Why?" 😄 

I took this logic from https://github.com/nasa/cumulus/pull/2245/files#diff-2e28b62d59bb9e81dcbe3b0f48929ad59c25e359b98cc9d67571d96c3e3123f5 because I think the idea is that we could have deleted the Dynamo record in the transaction and have the transaction still fail for _reasons_ after. That, or I'm misunderstanding how transactions fundamentally work and _both_ deletes will execute even if the first one fails. I don't think that's right but need to verify why we're re-creating the Dynamo record like this.

Jkovarik replied:
Wrote a bunch of commentary before seeing this comment ^      See above.

markdboyd replied:
The pattern of "recreate the record" if the transaction fails is not necessary for the standard PG/dynamo delete case. In that case, if either delete fails, the other doesn't happen.

The pattern **is necessary** once you add a third operation to the transaction, such as delete from Elasticsearch (which I'm doing as part of 2208). Because in that case you could have:

- delete from PG succeeds
- delete from Dynamo succeeds
- delete from Elasticsearch errors/throws

However, the Dynamo delete wouldn't "roll back" when the Elasticsearch delete errors/throws, so I added logic to manually recreate the Dynamo record in those cases. Otherwise the three systems would be out of sync. It's a bit unfortunate but a necessary evil while we are temporarily maintaining data across three systems and ultimately gives us a transactional guarantee about the state of data between them. 

So this pattern isn't necessary until we add writing to/deleting from Elasticsearch to the endpoint logic, which doesn't exist in this PR.

npauzenga replied:
AH! That makes sense. We were thinking there was a good reason but couldn't find it 😄.

It would make sense to me to leave this as-is on both the async operations PR and this one for consistency and to support ES down the line. Unless it creates duplicate Dynamo records 🤔. Is there a chance we could fail the PG delete, not attempt the Dynamo delete, and end up creating a duplicate Dynamo record in the `catch`? I could write a test to try it.

markdboyd replied:
I would suggest going the other way: remove the extra try/catch and recreation of the dynamo record. At best, it's confusing to leave it in when we don't need it yet. And at worst, it may have some unintended side effects if it's left in, because the Dynamo recreate will **overwrite** the non-deleted Dynamo record.

npauzenga replied:
Ah! Nope, I missed this. I think I saw the 👍 on Mark's response and thought I already hit it.

Here's the change to remove the try/catch and an added test for trying to delete an execution that doesn't exist. I think it makes sense to leave the Dynamo lookup before trying to delete so we can return that 404. Alternatively we might be able to look at the number of rows affected by the delete and if it's 0 say the record couldn't be found. I'm not totally sure that's a good pattern though. It might be possible that a record fails to delete due to a fk constraint and, in that case, we wouldn't want to return a 404.

https://github.com/nasa/cumulus/pull/2252/commits/91eebde2c710bb63f3d28d0829db592f2a3f0018

Jkovarik replied:
👍🏻  looks good.

Jkovarik commented:
(we should probably track a ticket for this)

Jkovarik commented on packages/api/endpoints/executions.js:
```diff
@@ -45,7 +46,50 @@ async function get(req, res) {
+    // Delete is idempotent, so there may not be a DynamoDB
+    // record to recreate
+    if (existingExecution) {
+      await executionModel.create(existingExecution);
```
It's correct that a failure to delete ..... due to lack of a record to delete would pass (returning 0 rows deleted), but that's not a failure we can catch sanely in this 'maybe it's migrated, maybe it's not' environment....  

Still, it's not clear to me how we have a failure scenario in the try/catch where the execution is *actually* deleted unless there's a rollback failure or some other internal transactional issue.    And if so, something like a thrown error on knex connect is going to attempt to do a model.create when the record hasn't been deleted.

Jkovarik replied:
I wrote a quick base model unit test for db/tests/models:

test.only('BasePgModel.delete() does the right thing with a transaction', async (t) => {
  const { knex, basePgModel, tableName } = t.context;
  const info = cryptoRandomString({ length: 5 });

  // Insert the records and validate that they exists in the table
  const [[recordCumulusId1], [recordCumulusId2]] = await Promise.all([
    knex(tableName)
      .insert({ info })
      .returning('cumulus_id'),
    knex(tableName)
      .insert({ info })
      .returning('cumulus_id'),
  ]);

  t.true(await basePgModel.exists(knex, { cumulus_id: recordCumulusId1 }));
  t.true(await basePgModel.exists(knex, { cumulus_id: recordCumulusId2 }));

  let result;
  try { 
    await knex.transaction(async (trx) => {
      await trx('someOtherTableName').where({}).del();
      result = 'Danger Will Robinson';
    });
  } catch (_e) {
    console.log('error');
  }
  t.is(result, undefined);

  t.true(await basePgModel.exists(knex, { cumulus_id: recordCumulusId1 }));
  t.true(await basePgModel.exists(knex, { cumulus_id: recordCumulusId2 }));
});


Based on ^, *at least for knex errors*, obviously 0 rows returned are different, I'm not sure this is doing what we want here.

Jkovarik replied:
^ you could reasonably do a t.throwsAsync for the try/catch function above, but I was just spitballing this for the kinds of cases we're using it for in the api endpoint.

npauzenga replied:
Closing this thread in favor of https://github.com/nasa/cumulus/pull/2252#discussion_r638777192

Jkovarik reviewed (changes requested):
Posting initial review comments, still reviewing!

Jkovarik commented on example/spec/parallel/syncGranule/SyncGranuleDuplicateHandlingSpec.js:
```diff
@@ -58,6 +59,12 @@ describe('When the Sync Granule workflow is configured', () => {
   let pdrModel;
   let testSuffix;
   let testDataFolder;
+  let syncGranuleExecutionArn;
```
If these are all top-scoped, consider an array  to iterate over instead.     

That said,I think we should  locally scope this if cleanup in a sub-test afterAll is where these are being used, and make the names locally meaningful.

npauzenga replied:
Yup, that's cleaner. Could also put all of these at the top-level in `afterAll` but I think it might be cleaner to leave them. This is a test that we could look at when we get a chance. It's creating a _lot_ of executions.

https://github.com/nasa/cumulus/pull/2252/commits/4ea600e788c2f7a2b16c785c26ed785620525611

Jkovarik commented on example/spec/parallel/kinesisTests/LambdaEventSourceSpec.js:
```diff
@@ -152,8 +156,9 @@ describe('When adding multiple rules that share a kinesis event stream', () => {
+      executionArn = workflowExecutions[0].executionArn;
 
-      const taskInput = await lambdaStep.getStepInput(workflowExecutions[0].executionArn, 'HelloWorld');
+      const taskInput = await lambdaStep.getStepInput(executionArn, 'HelloWorld');
```
👍🏻

Jkovarik commented on example/spec/parallel/kinesisTests/KinesisReplaySpec.js:
```diff
@@ -215,6 +216,10 @@ describe('The Kinesis Replay API', () => {
+
+        console.log('Cleaning up executions...');
+        await deleteExecution({ prefix: testConfig.stackName, executionArn: workflowExecutions[0].workflowArn });
+        await deleteExecution({ prefix: testConfig.stackName, executionArn: workflowExecutions[1].workflowArn });
```
Consider  a promise.allSettled / catch.   Possibly a DRY'd up helper function for these sorts of cleanups?

npauzenga replied:
Yup! These don't need to be in order. Also, I think this is one of the last errors I was looking for. I was using the wrong property here and it was throwing a 404. Updated here:

https://github.com/nasa/cumulus/pull/2252/commits/8106aed787b4b836b09d21cd6af8dca6f3a9e4d3

Jkovarik reviewed (changes requested):
Adding some additional comments re: tests.     Just a few more things to consider/discuss.  

This is really good work, and should help us get to the bottom of the integration test issues.    👍🏻

Jkovarik commented on example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js:
```diff
@@ -731,6 +736,8 @@ describe('When there are granule differences and granule reconciliation is run',
     console.log(`update granule files back ${publishedGranuleId}`);
     await granuleModel.update({ granuleId: publishedGranuleId }, { files: JSON.parse(granuleBeforeUpdate.body).files });
 
+    await deleteExecution({ prefix: config.stackName, executionArn: ingestGranuleExecutionArn });
```
This is majorly exciting.

npauzenga commented on packages/api/endpoints/executions.js:
```diff
@@ -45,7 +46,50 @@ async function get(req, res) {
+ * @param {Object} res - express response object
+ * @returns {Promise<Object>} the promise of express response object
+ */
+async function deleteExecution(req, res) {
```
👍 updated in one of the many many recent commits.

npauzenga commented on packages/api/tests/endpoints/executions.js:
```diff
@@ -123,3 +157,40 @@ test('GET executions with asyncOperationId filter returns the correct executions
   t.is(response.body.results[0].arn, 'arn2');
 });
+
+test('DELETE removes executions from all data stores', async (t) => {
```
🤔 not sure what you mean by
>we should probably make sure :arn can't be something that would resolve to in the db model

But here's an updated test that ensures we're ONLY deleting the record we specify:
https://github.com/nasa/cumulus/pull/2252/commits/f0610d8997707452a761fffae0e44fba3f1bfca9

Jkovarik replied:
Sorry, that got truncated somehow.     In either case deleting only the specified record addresses the concern.

npauzenga commented on example/spec/serial/DiscoverGranulesHttpSpec.js:
```diff
@@ -248,6 +262,8 @@ describe('The Discover Granules workflow with http Protocol', () => {
   });
 
   describe('the DiscoverGranules Lambda ignoring files config', () => {
+    let discoverGranulesExecutionArn3;
```
Yup, the tests that create a bunch of executions get really confusing fast. Here's the name change:
https://github.com/nasa/cumulus/pull/2252/commits/3e8022ec163a24a9b7916e905512f95bbbf4f60d

Totally hear ya on the scope of this test. Given the discussion during retro today it seems reasonable to take a close look at all of these tests.

npauzenga commented on example/spec/parallel/testAPI/ruleSpec.js:
```diff
@@ -248,7 +260,14 @@ describe('When I create a one-time rule via the Cumulus API', () => {
+    });
+
+    afterAll(async () => {
+      await deleteExecution({ prefix: config.stackName, executionArn: executionArn });
```
👍 I was thinking this needed to be in a set order but it appears not.

https://github.com/nasa/cumulus/pull/2252/commits/7f3aab42b474b31651f08304ef0d02972396d3eb

npauzenga commented on example/spec/parallel/syncGranule/SyncGranuleSuccessSpec.js:
```diff
@@ -334,15 +348,24 @@ describe('The Sync Granules workflow', () => {
 
+    afterAll(async () => {
+      // clean up stack state added by test
+      await deleteExecution({ prefix: config.stackName, executionArn: failingExecutionArn });
```
I'm not sure I'm totally following. I cleaned this up a bit and consolidated all of the execution cleanup in the parent `afterAll` here:

https://github.com/nasa/cumulus/pull/2252/commits/453d94bad7d0f1c119e9a50f8f3e297e80a783d0

I know that wasn't your point here. Do you mean that there's a chance the `failingExecution` won't resolve with an ARN? I haven't seen that but it's possible maybe?

On your larger point, yes. If we log any non-200 responses with the test name, perhaps that an intermediate step to get us better logs. Maybe we do that in all of our tests? Might do that here or in a separate PR.

Jkovarik commented on example/spec/parallel/ingestGranule/IngestUMMGSuccessSpec.js:
```diff
@@ -129,7 +129,6 @@ describe('The S3 Ingest Granules workflow configured to ingest UMM-G', () => {
 
     process.env.GranulesTable = `${config.stackName}-GranulesTable`;
     process.env.ExecutionsTable = `${config.stackName}-ExecutionsTable`;
-    executionModel = new Execution();
```
👍🏻     Just came across this on another ticket.   Nice work.

npauzenga commented on example/spec/parallel/ingestGranule/IngestGranuleSuccessSpec.js:
```diff
@@ -254,28 +257,31 @@ describe('The S3 Ingest Granules workflow', () => {
   afterAll(async () => {
     // clean up stack state added by test
-    await Promise.all([
+    const a = await apiTestUtils.deletePdr({
```
Temporary logging. Will remove once verified.

npauzenga commented on example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js:
```diff
@@ -741,6 +750,7 @@ describe('When there are granule differences and granule reconciliation is run',
       extraCumulusCollectionCleanup(),
       cmrClient.deleteGranule(cmrGranule),
     ]);
+    console.log('CreateReconcilliationReportSpec afterAll:::', a, b, c);
```
Will remove.

npauzenga replied:
(removed)

npauzenga commented on packages/integration-tests/index.js:
```diff
@@ -325,14 +325,14 @@ const loadCollection = async (params = {}) =>
   setProcessEnvironment(stackName, bucketName);
 
-  await Promise.all(
+  return await Promise.all(
```
This caused a lot of grief. Previously it was returning `collections.length;`, which was always the length of the input collections. It looked like it deleted the appropriate rows but didn't necessarily.

Jkovarik replied:
:thumbsup: nice catch.

Jkovarik reviewed (changes requested):
Really good work - there's a couple of bits of cleanup and the commentary around the deletion logic is still outstanding.

npauzenga commented on example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js:
```diff
@@ -731,7 +737,10 @@ describe('When there are granule differences and granule reconciliation is run',
     await granuleModel.update({ granuleId: publishedGranuleId }, { files: JSON.parse(granuleBeforeUpdate.body).files });
 
-    await Promise.all([
+    const a = await deleteExecution({ prefix: config.stackName, executionArn: ingestGranuleExecutionArn });
```
Yeah, I'll remove all of the logging before merge.

npauzenga replied:
(removed)

npauzenga commented on example/spec/parallel/testAPI/sqsRuleSpec.js:
```diff
@@ -80,19 +84,30 @@ async function cleanUp() {
     deleteQueue(queues.deadLetterQueueUrl),
   ]);
+
+  console.log('sqsRuleSpec afterAll:::', x);
```
Yup, should have pointed this out. There are a couple tests with this kind of logging that I committed to run in CI due to frustration 😄 . I'm going to go through everything and remove logging before merging.

npauzenga commented on example/spec/serial/DiscoverGranulesHttpSpec.js:
```diff
@@ -84,24 +93,35 @@ describe('The Discover Granules workflow with http Protocol', () => {
       cleanupCollections(config.stackName, config.bucket, collectionsDir, testSuffix),
       deleteProvider({ prefix: config.stackName, providerId: provider.id }),
     ]);
+    console.log('DiscoverGranulesHTTPSpec afterAll:::', a, b, c, d, e, f, g, x);
```
yup, same comment. Will remove all logging. I just couldn't get this working locally so needed to see it in CI.

npauzenga commented on example/spec/parallel/discoverGranules/DiscoverGranulesHttpsSpec.js:
```diff
@@ -135,23 +147,36 @@ describe('The Discover Granules workflow with https Protocol', () => {
-      ingestGranuleWorkflowArn = queueGranulesOutput.payload.running[0];
-      console.log('\nwait for ingestGranuleWorkflow', ingestGranuleWorkflowArn);
-      ingestGranuleExecutionStatus = await waitForCompletedExecution(ingestGranuleWorkflowArn);
+      ingestGranuleWorkflowArn1 = queueGranulesOutput.payload.running[0];
```
👍 https://github.com/nasa/cumulus/pull/2252/commits/fbbc616dd846e5de0d5b45855acf0c7c7014a60b

Jkovarik replied:
Nice, thanks for addressing that!

npauzenga commented on example/spec/parallel/discoverGranules/DiscoverGranulesS3FailureSpec.js:
```diff
@@ -65,15 +66,17 @@ describe('The DiscoverGranules workflow with a non-existent bucket', () => {
-    Promise.all([
+  afterAll(async () => {
+    await deleteExecution({ prefix: stackName, executionArn: workflowExecution.executionArn });
+    await Promise.all([
```
Yup, that would be great. That helper could also maybe help with logging.

npauzenga commented on example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingErrorSpec.js:
```diff
@@ -25,19 +26,22 @@ const { loadConfig } = require('../../helpers/testUtils');
   let collection;
+  let config;
   let firstIngestGranuleRule;
+  let firstIngestGranuleExecutionArn;
```
Yeah, I should have updated these as I went. Instead I just did whichever format was in the test. Here's an update to use numerals instead of `first`/`second`. 

https://github.com/nasa/cumulus/pull/2252/commits/3341ede4e9d6bb4a676754070fa558a4e2ff22c3

npauzenga commented on example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingVersionSpec.js:
```diff
@@ -25,9 +26,11 @@ const { loadConfig } = require('../../helpers/testUtils');
   let differentChecksumFilename;
   let differentChecksumKey;
   let firstIngestGranuleRule;
+  let firstIngestGranuleExecutionArn;
```
👍 

https://github.com/nasa/cumulus/pull/2252/commits/3341ede4e9d6bb4a676754070fa558a4e2ff22c3

npauzenga commented on example/spec/parallel/retryConfiguration/retryConfigurationSpec.js:
```diff
@@ -24,6 +27,8 @@ function getRetryIntervals(executions) {
   return retryIntervals;
 }
 
+// TODO delete executions
```
👍 https://github.com/nasa/cumulus/pull/2252/commits/2def3543ee743216f3482bfe1697dfbf1bedff7a

npauzenga commented on packages/api/tests/endpoints/executions.js:
```diff
@@ -115,11 +150,68 @@ test('GET logs returns all executions', async (t) => {
+  executions.push(newExecution);
+
+  await Promise.all(executions.map(async (execution) => {
+    delete execution.asyncOperationId;
```
Oh so the reason I'm doing this is to avoid needing to create an Async Operation just to test execution deletion. If we have that `execution.asyncOperationId` and I go to create the PG record, I'll need to create a valid async operation with a matching ID. That isn't necessary for testing execution deletion here so I'm removing it.

Added a comment for clarification:
https://github.com/nasa/cumulus/pull/2252/commits/e886c0ce468fa396de61f53db59ec96963f73610

Jkovarik replied:
Oops, no, that's my bad, I misread that as explicitly deleting the *record* not the object key.  🤦🏻        Thanks for the comment, in either case 👍🏻

npauzenga commented on example/spec/serial/DiscoverGranulesHttpSpec.js:
```diff
@@ -94,20 +97,23 @@ describe('The Discover Granules workflow with http Protocol', () => {
+
+    await deleteExecution({ prefix: config.stackName, executionArn: ignoringFilesConfigExecutionArn });
+    await deleteExecution({ prefix: config.stackName, executionArn: partialFilesConfigExecutionArn });
+    await deleteExecution({ prefix: config.stackName, executionArn: noFilesConfigExecutionArn });
```
This test created 16 executions!

npauzenga commented:
Removed logging and addressed the actionable comments above. Running in CI now. I expect there still may be a couple errors in the logs but it's down to just a couple. I can definitely get this to 100% but I'm a bit hesitant to sink more time into it (it takes a long time to track these things down) if we're thinking we'll be able to address some of our test instability in the IP. Don't want to leave things hanging if not though so this can be a discussion.

Jkovarik commented on example/spec/parallel/testAPI/ruleSpec.js:
```diff
@@ -248,7 +260,16 @@ describe('When I create a one-time rule via the Cumulus API', () => {
+
+    afterAll(async () => {
+      await Promise.allSettled([
+        deleteExecution({ prefix: config.stackName, executionArn: executionArn }),
```
Oops - commented on the commit, re-posting: 

We should probably still be checking for a  rejected promise in the array here and logging/throwing/etc, or it'll be *really* hard to tell what's going on if it fails.

Jkovarik replied:
👍🏻  to switch back to `.all`, in lieu of fixing this more broadly in the integration tests.

Jkovarik commented on example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js:
```diff
@@ -731,6 +737,10 @@ describe('When there are granule differences and granule reconciliation is run',
     console.log(`update granule files back ${publishedGranuleId}`);
     await granuleModel.update({ granuleId: publishedGranuleId }, { files: JSON.parse(granuleBeforeUpdate.body).files });
 
+    // TODO there may be another execution to delete here
```
Is this an errant comment, or is it something we're intending to clean up in another ticket?

npauzenga replied:
Yeah, this is one I'm not sure of. There are 2 specs with this comment. If there are unaccounted for execution errors these were going to be the first places to look. This may be moot with your previous comment though.

Jkovarik reviewed (approved):
Great work, thanks for addressing the comments, etc.
